File: src/main/java/medium/MaskingPersonallyIdentifiableInformation.java
package medium;

public class MaskingPersonallyIdentifiableInformation {
  public String maskPhone(String phone) {
    StringBuilder sb = new StringBuilder();
    int count = 0;
    int chars = 0;
    for (int i = phone.length() - 1; i >= 0; i--) {
      char c = phone.charAt(i);
      if (!Character.isDigit(c)) continue;
      if (count == 4 || count == 8 || count == 12) {
        count++;
        sb.append("-");
      }
      if (count <= 4) {
        sb.append(c);
      } else {
        sb.append("*");
      }
      count++;
      chars++;
    }
    if (chars > 10) {
      sb.append("+");
    }
    sb.reverse();
    return sb.toString();
  }

  public String maskEmail(String email) {
    StringBuilder sb = new StringBuilder();
    int atIndex = email.indexOf("@");
    sb.append(email.charAt(0));
    sb.append("*****");
    sb.append(email.charAt(atIndex - 1));
    sb.append(email.substring(atIndex));
    return sb.toString().toLowerCase();
  }

  public String solve(String s) {
    if (s.contains("@")) return maskEmail(s);
    return maskPhone(s);
  }
}

File: src/medium/masking-personally-identifiable-information.ts
export const maskEmail = (s: string): string => {
	const result: string[] = []
	const domain = s.indexOf("@")
	for(let i = 0; i < s.length; i++) {
		if(i === 0) {
			result.push(s[i], "*****")
		}else if(i === domain - 1 || i >= domain) {
			result.push(s[i])
		}
	}
		return result.join("")
}
export const maskPhone = (s: string): string => {
	const result: string[] = []
	let count = 0;
	let chars = 0;
	for(let i = s.length - 1; i >= 0; i--) {
		if(s[i].charCodeAt(0) < 48 || s[i].charCodeAt(0) > 57) continue;
		if(count === 4|| count === 8 || count === 12) {
			result.unshift("-")
			count++;
		}
		if(count <= 4) {
			result.unshift(s[i])
		}else {
			result.unshift("*")
		}
		count++
		chars++
	}
	if(chars > 10) {
		result.unshift("+")
	}
	return result.join("")
}
export const maskingPersonallyIdentifiableInformation = (s: string): string => (s.includes("@")) ? maskEmail(s.toLowerCase()) : maskPhone(s)
